### ISSUE TITLE
SUPG for Quadruped Locomotion - 4-legged Walker

### ISSUE THREAD

**Issue by MuleaneEve** [colgreen/sharpneat#18]
Last year, I was learning about ANN and EA by playing with existing projects, and I particularly liked this 3D 4-legged walker based on a [research paper by Gregory Morse](http://eplex.cs.ucf.edu/supg/supgql/). You can watch videos of the walker on his website.

So, I updated the source code a bit, and [uploaded it on bitbucket](https://bitbucket.org/MuleaneEve/ode.net/src/79dc3ee1ecbf5dc9a173e3f8a34b05041e8a0c78/OscillatorQuadruped/?at=default) (because of my other interest in physics simulation).

I am mentioning it now after reading the issue #16 by @asimaranov and @drallensmith.
Sadly, right now, it is using a very old version of SharpNeatLib and I don't have enough mastery to update it to the current version of sharpneat.

Would you be interested in it?
I would help as much as I can.

**Comment by colgreen** [comment]
Thanks. Potentially this could become one of the default task/domains following a review of the code. However, right now my time is directed towards a big refactor of the core code so I won't have time to do anything for a while.

As such I will leave this open for the time being.

**Comment by MuleaneEve** [comment]
Thank you. I look forward to both.

**Comment by MuleaneEve** [comment]
I will re-write this sample to use [bepuphysics](https://github.com/bepu/bepuphysics2) instead of ODE.

The benefits are:
- bepuphysics is written in pure C#, unlike ODE which is in C++.
- It is under active development, unlike ODE which is mostly dormant now. And it's hosted on GitHub.
- We will benefit from "cross-pollination": Developers using bepuphysics will discover SharpNEAT more easily and may start using it, and vice-versa.

The only downside that I see is that bepuphysics v2 was just released in alpha, so it is less stable than ODE. However, that should be a short term inconvenience and this sample doesn't require complex physics.

**Comment by colgreen** [comment]
Yeh looks good, and certainly seems fast based on how much is going on in some of the demos - and those include the rendering overhead too (seems to be using a managed library over DirectX).

Did you mean you will re-write your code currently on drop box? I.e. should I wait for this before bringing the work into sharpneat?

**Comment by KPixel** [comment]
Yes, I will re-write the code (from bitbucket) and host it in a temporary repository on GitHub.
You can then take it into SharpNEAT.

By the way, my expectation is that the quadruplet will no longer be able to learn to walk with BEPUphysics. The old code is slow to learn (it can take over 5 hours) and very brittle (most of the time, the learning fails and you have to start over).

So, my focus will be to set up the project with the 3D and physics aspects, so that you can take care of the AI.

**Comment by MuleaneEve** [comment]
Yes, I will re-write the code (from bitbucket) and host it in a temporary repository on GitHub.
You can then take it into SharpNEAT.

By the way, my expectation is that the quadruplet will no longer be able to learn to walk with BEPUphysics. The old code is slow to learn (it can take over 5 hours) and very brittle (most of the time, the learning fails and you have to start over).

So, my focus will be to set up the project with the 3D and physics aspects, so that you can take care of the AI.

**Comment by MuleaneEve** [comment]
I have made some progress on this project. [The code is available here](https://github.com/MuleaneEve/bepuphysics2).
I also [introduced it to the BEPUphysics devs](https://github.com/bepu/bepuphysics2/issues/53).

The Quadruped is currently just a simple rag-doll. It hasn't been tied to SharpNEAT yet. The code to achieve that is in [SharpNeatWalker\OscillatorQuadruped](https://github.com/MuleaneEve/bepuphysics2/tree/master/SharpNeatWalker/OscillatorQuadruped) but it is commented out because it needs to be adapted to this new engine.

One idea that I have for this new version is that the app can show the best neural network running on screen, while the training is ongoing in the background. And we can introduce some input keys to control all that...
